README.md
[![NPM version](https://img.shields.io/npm/v/@gmod/bam.svg?style=flat-square)](https://npmjs.org/package/@gmod/bam)
[![Coverage Status](https://img.shields.io/codecov/c/github/GMOD/bam-js/master.svg?style=flat-square)](https://codecov.io/gh/GMOD/bam-js/branch/master)
[![Build Status](https://img.shields.io/github/actions/workflow/status/GMOD/bam-js/push.yml?branch=master)](https://github.com/GMOD/bam-js/actions?query=branch%3Amaster+workflow%3APush+)

## Install

```bash
$ npm install --save @gmod/bam
```

## Usage

```typescript
const { BamFile } = require('@gmod/bam')
// or import {BamFile} from '@gmod/bam'

const t = new BamFile({
  bamPath: 'test.bam',
})

// note: it's required to first run getHeader before any getRecordsForRange
var header = await t.getHeader()

// this would get same records as samtools view ctgA:1-50000
var records = await t.getRecordsForRange('ctgA', 0, 50000)
```

The `bamPath` argument only works on nodejs. In the browser, you should pass
`bamFilehandle` with a generic-filehandle2 e.g. `RemoteFile`

```typescript
const { RemoteFile } = require('generic-filehandle2')
const bam = new BamFile({
  bamFilehandle: new RemoteFile('yourfile.bam'), // or a full http url
  baiFilehandle: new RemoteFile('yourfile.bam.bai'), // or a full http url
})
```

Input are 0-based half-open coordinates (note: not the same as samtools view
coordinate inputs!)

## Usage with htsget

Since 1.0.41 we support usage of the htsget protocol

Here is a small code snippet for this

```typescript
const { HtsgetFile } = require('@gmod/bam')

const ti = new HtsgetFile({
  baseUrl: 'http://htsnexus.rnd.dnanex.us/v1/reads',
  trackId: 'BroadHiSeqX_b37/NA12878',
})
await ti.getHeader()
const records = await ti.getRecordsForRange(1, 2000000, 2000001)
```

Our implementation makes some assumptions about how the protocol is implemented,
so let us know if it doesn't work for your use case

## Documentation

### BAM constructor

The BAM class constructor accepts arguments

- `bamPath`/`bamUrl`/`bamFilehandle` - a string file path to a local file or a
  class object with a read method
- `csiPath`/`csiUrl`/`csiFilehandle` - a CSI index for the BAM file, required
  for long chromosomes greater than 2^29 in length
- `baiPath`/`baiUrl`/`baiFilehandle` - a BAI index for the BAM file
- `cacheSize` - limit on number of chunks to cache. default: 50
- `yieldThreadTime` - the interval at which the code yields to the main thread
  when it is parsing a lot of data. default: 100ms. Set to 0 to performed no
  yielding

Note: filehandles implement the Filehandle interface from
https://www.npmjs.com/package/generic-filehandle2. This module offers the path
and url arguments as convenience methods for supplying the LocalFile and
RemoteFile

### async getRecordsForRange(refName, start, end, opts)

Note: you must run getHeader before running getRecordsForRange

- `refName` - a string for the chrom to fetch from
- `start` - a 0-based half open start coordinate
- `end` - a 0-based half open end coordinate
- `opts.signal` - an AbortSignal to indicate stop processing
- `opts.viewAsPairs` - re-dispatches requests to find mate pairs. default: false
- `opts.pairAcrossChr` - control the viewAsPairs option behavior to pair across
  chromosomes. default: false
- `opts.maxInsertSize` - control the viewAsPairs option behavior to limit
  distance within a chromosome to fetch. default: 200kb

### async \*streamRecordsForRange(refName, start, end, opts)

This is a async generator function that takes the same signature as
`getRecordsForRange` but results can be processed using

```typescript
for await (const chunk of file.streamRecordsForRange(
  refName,
  start,
  end,
  opts,
)) {
}
```

The `getRecordsForRange` simply wraps this process by concatenating chunks into
an array

### async getHeader(opts: {....anything to pass to generic-filehandle2 opts})

This obtains the header from `HtsgetFile` or `BamFile`. Retrieves BAM file and
BAI/CSI header if applicable, or API request for refnames from htsget

### async indexCov(refName, start, end)

- `refName` - a string for the chrom to fetch from
- `start` - a 0-based half open start coordinate (optional)
- `end` - a 0-based half open end coordinate (optional)

Returns features of the form {start, end, score} containing estimated feature
density across 16kb windows in the genome

### async lineCount(refName: string)

- `refName` - a string for the chrom to fetch from

Returns number of features on refName, uses special pseudo-bin from the BAI/CSI
index (e.g. bin 37450 from bai, returning n_mapped from SAM spec pdf) or -1 if
refName not exist in sample

### async hasRefSeq(refName: string)

- `refName` - a string for the chrom to check

Returns whether we have this refName in the sample

### Returned features

The returned features from BAM are lazy features meaning that it delays
processing of all the feature tags until necessary.

You can access data feature.get('field') to get the value of a feature attribute

Example

```typescript
feature.get('seq_id') // numerical sequence id corresponding to position in the sam header
feature.get('start') // 0-based half open start coordinate
feature.get('end') // 0-based half open end coordinate
```

#### Fields

```typescript
feature.get('name') // QNAME
feature.get('seq') // feature sequence
feature.get('qual') // qualities
feature.get('cigar') // cigar string
feature.get('MD') // MD string
feature.get('SA') // supplementary alignments
feature.get('template_length') // TLEN
feature.get('length_on_ref') // derived from CIGAR using standard algorithm
```

#### Flags

```typescript
feature.get('flags') // see https://broadinstitute.github.io/picard/explain-flags.html
```

#### Tags

BAM tags such as MD can be obtained via

```typescript
feature.get('MD')
```

A full list of tags that can be obtained can be obtained via

    feature._tags()

The feature format may change in future versions to be more raw data records,
but this will be a major version bump

#### Note

The reason that we hide the data behind this ".get" function is that we lazily
decode records on demand, which can reduce memory consumption.

## License

MIT © [Colin Diesh](https://github.com/cmdcolin)


CHANGELOG.md
# [2.0.0](https://github.com/GMOD/bam-js/compare/v1.1.18...v2.0.0) (2023-06-08)


### Features

* explicit buffer import ([#98](https://github.com/GMOD/bam-js/issues/98)) ([66de9f4](https://github.com/GMOD/bam-js/commit/66de9f4ce30e3ff647d5297f093695e92ec9227c))



- Add explicit buffer import
- Remove cross-fetch and object.entries polyfills
- Improve typescripting
- Remove chunkSizeLimit and fetchSizeLimit

## [1.1.18](https://github.com/GMOD/bam-js/compare/v1.1.17...v1.1.18) (2022-12-17)

- Use es2015 for nodejs build

## [1.1.17](https://github.com/GMOD/bam-js/compare/v1.1.16...v1.1.17) (2022-07-18)

- Bump devDeps and generic-filehandle2 to 3.0.0

## [1.1.16](https://github.com/GMOD/bam-js/compare/v1.1.15...v1.1.16) (2022-03-30)

- Add src directory for better source maps

## [1.1.15](https://github.com/GMOD/bam-js/compare/v1.1.14...v1.1.15) (2022-03-18)

- Fix for htsget failing with message 'input must be buffer, number, or string,
  received object'
- Speed improvement by caching chunks of features

## [1.1.14](https://github.com/GMOD/bam-js/compare/v1.1.13...v1.1.14) (2022-03-14)

- Fix seq function for corner case

## [1.1.13](https://github.com/GMOD/bam-js/compare/v1.1.12...v1.1.13) (2022-02-25)

- Optimize qual and sequence string record functions for less GC pressure

<a name="1.1.12"></a>

## [1.1.12](https://github.com/GMOD/bam-js/compare/v1.1.11...v1.1.12) (2022-02-17)

- Add blocksForRange method to BamFile class to help stats estimation in JBrowse
  2

<a name="1.1.11"></a>

## [1.1.11](https://github.com/GMOD/bam-js/compare/v1.1.10...v1.1.11) (2022-01-26)

- Cache setup of index file parsing

<a name="1.1.10"></a>

## [1.1.10](https://github.com/GMOD/bam-js/compare/v1.1.9...v1.1.10) (2022-01-18)

- Make \_refID and flags public fields
- Small internal changes to the handling of opts

<a name="1.1.9"></a>

## [1.1.9](https://github.com/GMOD/bam-js/compare/v1.1.8...v1.1.9) (2021-12-14)

- Add ESM module export in package.json (smaller bundle size for consumers)
- Cache BAI readFile result for compatibility with node.js native filehandles
  (which otherwise fail if re-reading the filehandle twice)

<a name="1.1.8"></a>

## [1.1.8](https://github.com/GMOD/bam-js/compare/v1.1.7...v1.1.8) (2021-05-21)

- Fix types for yieldThreadTime

<a name="1.1.7"></a>

## [1.1.7](https://github.com/GMOD/bam-js/compare/v1.1.6...v1.1.7) (2021-05-21)

- New param yieldThreadTime to constructor to yield while processing

<a name="1.1.6"></a>

## [1.1.6](https://github.com/GMOD/bam-js/compare/v1.1.5...v1.1.6) (2021-02-20)

- Add qualRaw function on records for getting raw qual score array instead of
  string

<a name="1.1.5"></a>

## [1.1.5](https://github.com/GMOD/bam-js/compare/v1.1.4...v1.1.5) (2020-12-11)

- Allow getHeaderText to accept cancellation options

<a name="1.1.4"></a>

## [1.1.4](https://github.com/GMOD/bam-js/compare/v1.1.3...v1.1.4) (2020-12-11)

- Add canMergeBlocks to CSI code (already existed in BAI)
- Add suggestion from @jrobinso about reg2bins modification for memory saving
  (Thanks!)
- Add getHeaderText() method for getting a text string of the header data

<a name="1.1.3"></a>

## [1.1.3](https://github.com/GMOD/bam-js/compare/v1.1.2...v1.1.3) (2020-10-29)

- Fix usage of feature.get('seq'), was using feature.getReadBases before this

<a name="1.1.2"></a>

## [1.1.2](https://github.com/GMOD/bam-js/compare/v1.1.1...v1.1.2) (2020-10-02)

- Fix signedness in BAM tags (#65)
- Remove unused seq_reverse_complemented tag from \_tags()

<a name="1.1.1"></a>

## [1.1.1](https://github.com/GMOD/bam-js/compare/v1.1.0...v1.1.1) (2020-09-20)

- Remove JBrowse specific results from tags

<a name="1.1.0"></a>

# [1.1.0](https://github.com/GMOD/bam-js/compare/v1.0.42...v1.1.0) (2020-08-28)

- Add support for the CG tag for long CIGAR strings

<a name="1.0.42"></a>

## [1.0.42](https://github.com/GMOD/bam-js/compare/v1.0.41...v1.0.42) (2020-08-19)

- Small bugfix for Htsget specifically

<a name="1.0.41"></a>

## [1.0.41](https://github.com/GMOD/bam-js/compare/v1.0.40...v1.0.41) (2020-08-19)

- Add htsget example
- Support opts object to getHeader allowing things like auth headers to be
  passed right off the bat

<a name="1.0.40"></a>

## [1.0.40](https://github.com/GMOD/bam-js/compare/v1.0.39...v1.0.40) (2020-07-30)

<a name="1.0.39"></a>

## [1.0.39](https://github.com/GMOD/bam-js/compare/v1.0.38...v1.0.39) (2020-07-30)

- Don't use origin master in the follow-tags postpublish command for cleaner
  version publishing

<a name="1.0.38"></a>

## [1.0.38](https://github.com/GMOD/bam-js/compare/v1.0.37...v1.0.38) (2020-07-30)

- Direct construction of qual/seq toString
- Improve performance of the uniqueID calculation for pathological cases where
  there are tons of bins

<a name="1.0.37"></a>

## [1.0.37](https://github.com/GMOD/bam-js/compare/v1.0.36...v1.0.37) (2020-06-06)

- Typescript only release: export BamRecord types

<a name="1.0.36"></a>

## [1.0.36](https://github.com/GMOD/bam-js/compare/v1.0.35...v1.0.36) (2020-03-05)

- Adds a shortcut to stop parsing chunks after a record is detected to be
  outside the requested range while decoding

<a name="1.0.35"></a>

## [1.0.35](https://github.com/GMOD/bam-js/compare/v1.0.34...v1.0.35) (2020-02-04)

- Update scheme used to calculate unique fileOffset based IDs using
  @gmod/bgzf-filehandle updates

<a name="1.0.34"></a>

## [1.0.34](https://github.com/GMOD/bam-js/compare/v1.0.33...v1.0.34) (2020-01-24)

- Small fix for using id() instead of .get('id') for weird SAM records
  containing ID field

<a name="1.0.33"></a>

## [1.0.33](https://github.com/GMOD/bam-js/compare/v1.0.32...v1.0.33) (2020-01-24)

- Perform decoding of entire chunk up front to aid caching, reverts change in
  1.0.29

<a name="1.0.32"></a>

## [1.0.32](https://github.com/GMOD/bam-js/compare/v1.0.31...v1.0.32) (2019-11-16)

- Add a speed improvement for long reads by pre-allocating sequence/quality
  scores array

<a name="1.0.31"></a>


test/vanilla-node.test.ts
import fs from 'fs'
import { BAI, BamFile } from '../src'

xdescribe('using vanilla node filehandles', () => {
  it('loads volvox-sorted.bam.bai', async () => {
    const filehandle = await fs.promises.open('test/data/volvox-sorted.bam.bai')

    const ti = new BAI({
      // @ts-expect-error
      filehandle,
    })
    const indexData = await ti.parse()
    expect(indexData.bai).toEqual(true)
    expect(await ti.lineCount(0)).toEqual(9596)
    expect(await ti.hasRefSeq(0)).toEqual(true)
    filehandle.close()
  })

  it('gets features from volvox-sorted.bam', async () => {
    const bam = await fs.promises.open('test/data/volvox-sorted.bam')
    const bai = await fs.promises.open('test/data/volvox-sorted.bam.bai')
    const ti = new BamFile({
      // @ts-expect-error
      bamFilehandle: bam,
      // @ts-expect-error
      baiFilehandle: bai,
    })
    await ti.getHeader()
    const records = await ti.getRecordsForRange('ctgA', 0, 1000)
    expect(records.length).toEqual(131)
    expect(records[0].get('start')).toEqual(2)
    expect(records[0].get('end')).toEqual(102)
    expect(records[0].get('cigar')).toEqual('100M')
    expect(records[0].get('name')).toEqual('ctgA_3_555_0:0:0_2:0:0_102d')
    expect(records[0].get('qual')).toEqual(
      '17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17 17',
    )
    expect(records[0].get('md')).toEqual('100')
    expect(records[0].getReadBases()).toEqual(
      'TTGTTGCGGAGTTGAACAACGGCATTAGGAACACTTCCGTCTCTCACTTTTATACGATTATGATTGGTTCTTTAGCCTTGGTTTAGATTGGTAGTAGTAG',
    )
    bam.close()
    bai.close()
  })
})


src/util.ts
import Long from 'long'
import Chunk from './chunk'
import VirtualOffset from './virtualOffset'

export function timeout(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

export function longToNumber(long: Long) {
  if (
    long.greaterThan(Number.MAX_SAFE_INTEGER) ||
    long.lessThan(Number.MIN_SAFE_INTEGER)
  ) {
    throw new Error('integer overflow')
  }
  return long.toNumber()
}

/**
 * Properly check if the given AbortSignal is aborted.
 * Per the standard, if the signal reads as aborted,
 * this function throws either a DOMException AbortError, or a regular error
 * with a `code` attribute set to `ERR_ABORTED`.
 *
 * For convenience, passing `undefined` is a no-op
 *
 * @param {AbortSignal} [signal] an AbortSignal, or anything with an `aborted` attribute
 * @returns nothing
 */
export function checkAbortSignal(signal?: AbortSignal) {
  if (!signal) {
    return
  }

  if (signal.aborted) {
    // console.log('bam aborted!')
    if (typeof DOMException === 'undefined') {
      const e = new Error('aborted')
      //@ts-ignore
      e.code = 'ERR_ABORTED'
      throw e
    } else {
      throw new DOMException('aborted', 'AbortError')
    }
  }
}

/**
 * Skips to the next tick, then runs `checkAbortSignal`.
 * Await this to inside an otherwise synchronous loop to
 * provide a place to break when an abort signal is received.
 * @param {AbortSignal} signal
 */
export async function abortBreakPoint(signal?: AbortSignal) {
  await Promise.resolve()
  checkAbortSignal(signal)
}

export function canMergeBlocks(chunk1: Chunk, chunk2: Chunk) {
  return (
    chunk2.minv.blockPosition - chunk1.maxv.blockPosition < 65000 &&
    chunk2.maxv.blockPosition - chunk1.minv.blockPosition < 5000000
  )
}

export interface BamOpts {
  viewAsPairs?: boolean
  pairAcrossChr?: boolean
  maxInsertSize?: number
  signal?: AbortSignal
}

export interface BaseOpts {
  signal?: AbortSignal
}

export function makeOpts(obj: AbortSignal | BaseOpts = {}): BaseOpts {
  return 'aborted' in obj ? ({ signal: obj } as BaseOpts) : (obj as BaseOpts)
}

export function optimizeChunks(chunks: Chunk[], lowest?: VirtualOffset) {
  const mergedChunks: Chunk[] = []
  let lastChunk: Chunk | undefined

  if (chunks.length === 0) {
    return chunks
  }

  chunks.sort((c0, c1) => {
    const dif = c0.minv.blockPosition - c1.minv.blockPosition
    return dif === 0 ? c0.minv.dataPosition - c1.minv.dataPosition : dif
  })

  for (const chunk of chunks) {
    if (!lowest || chunk.maxv.compareTo(lowest) > 0) {
      if (lastChunk === undefined) {
        mergedChunks.push(chunk)
        lastChunk = chunk
      } else {
        if (canMergeBlocks(lastChunk, chunk)) {
          if (chunk.maxv.compareTo(lastChunk.maxv) > 0) {
            lastChunk.maxv = chunk.maxv
          }
        } else {
          mergedChunks.push(chunk)
          lastChunk = chunk
        }
      }
    }
  }

  return mergedChunks
}

export function parsePseudoBin(bytes: Uint8Array, offset: number) {
  const lineCount = longToNumber(
    Long.fromBytesLE(
      Array.prototype.slice.call(bytes, offset, offset + 8),
      true,
    ),
  )
  return { lineCount }
}

export function findFirstData(
  firstDataLine: VirtualOffset | undefined,
  virtualOffset: VirtualOffset,
) {
  return firstDataLine
    ? firstDataLine.compareTo(virtualOffset) > 0
      ? virtualOffset
      : firstDataLine
    : virtualOffset
}

export function parseNameBytes(
  namesBytes: Uint8Array,
  renameRefSeq: (arg: string) => string = s => s,
) {
  let currRefId = 0
  let currNameStart = 0
  const refIdToName = []
  const refNameToId: { [key: string]: number } = {}
  const decoder = new TextDecoder('utf8')
  for (let i = 0; i < namesBytes.length; i += 1) {
    if (!namesBytes[i]) {
      if (currNameStart < i) {
        let refName = decoder.decode(namesBytes.subarray(currNameStart, i))
        refName = renameRefSeq(refName)
        refIdToName[currRefId] = refName
        refNameToId[refName] = currRefId
      }
      currNameStart = i + 1
      currRefId += 1
    }
  }
  return { refNameToId, refIdToName }
}


src/record.ts
/* eslint-disable @typescript-eslint/no-empty-function */
import Constants from './constants'

const SEQRET_DECODER = '=ACMGRSVTWYHKDBN'.split('')
const CIGAR_DECODER = 'MIDNSHP=X???????'.split('')

/**
 * Class of each BAM record returned by this API.
 */
export default class BamRecord {
  private data = {} as { [key: string]: any }
  private bytes: { start: number; end: number; byteArray: Uint8Array }
  private dv: DataView
  private _id: number
  private _tagOffset: number | undefined
  private _tagList: string[] = []
  private _allTagsParsed = false

  public flags: any
  public _refID: number
  constructor(args: any) {
    const { bytes, fileOffset } = args
    const { start } = bytes
    this.data = {}
    this.bytes = bytes
    this._id = fileOffset
    this.dv = new DataView(
      this.bytes.byteArray.buffer,
      this.bytes.byteArray.byteOffset,
      this.bytes.byteArray.byteLength,
    )
    this._refID = this.dv.getInt32(start + 4, true)
    this.data.start = this.dv.getInt32(start + 8, true)
    this.flags = (this.dv.getInt32(start + 16, true) & 0xffff0000) >> 16
  }

  get(field: string) {
    //@ts-expect-error
    if (this[field]) {
      if (this.data[field]) {
        return this.data[field]
      }
      //@ts-expect-error
      this.data[field] = this[field]()
      return this.data[field]
    }
    return this._get(field.toLowerCase())
  }

  end() {
    return this.get('start') + this.get('length_on_ref')
  }

  seq_id() {
    return this._refID
  }

  // same as get(), except requires lower-case arguments.  used
  // internally to save lots of calls to field.toLowerCase()
  _get(field: string) {
    if (field in this.data) {
      return this.data[field]
    }
    this.data[field] = this._parseTag(field)
    return this.data[field]
  }

  _tags() {
    this._parseAllTags()

    let tags = ['seq']

    if (!this.isSegmentUnmapped()) {
      tags.push(
        'start',
        'end',
        'strand',
        'score',
        'qual',
        'MQ',
        'CIGAR',
        'length_on_ref',
        'template_length',
      )
    }
    if (this.isPaired()) {
      tags.push('next_segment_position', 'pair_orientation')
    }
    tags = tags.concat(this._tagList || [])

    for (const k of Object.keys(this.data)) {
      if (k[0] !== '_' && k !== 'next_seq_id') {
        tags.push(k)
      }
    }

    const seen: { [key: string]: boolean } = {}
    return tags.filter(t => {
      if (
        (t in this.data && this.data[t] === undefined) ||
        t === 'CG' ||
        t === 'cg'
      ) {
        return false
      }

      const lt = t.toLowerCase()
      const s = seen[lt]
      seen[lt] = true
      return !s
    })
  }

  parent() {
    return
  }

  children() {
    return this.get('subfeatures')
  }

  id() {
    return this._id
  }

  // special parsers
  /**
   * Mapping quality score.
   */
  mq() {
    const mq = (this.get('_bin_mq_nl') & 0xff00) >> 8
    return mq === 255 ? undefined : mq
  }

  score() {
    return this.get('mq')
  }

  qual() {
    return this.qualRaw()?.join(' ')
  }

  qualRaw() {
    if (this.isSegmentUnmapped()) {
      return
    }

    const { start, byteArray } = this.bytes
    const p =
      start +
      36 +
      this.get('_l_read_name') +
      this.get('_n_cigar_op') * 4 +
      this.get('_seq_bytes')
    const lseq = this.get('seq_length')
    return byteArray.subarray(p, p + lseq)
  }

  strand() {
    return this.isReverseComplemented() ? -1 : 1
  }

  multi_segment_next_segment_strand() {
    if (this.isMateUnmapped()) {
      return
    }
    return this.isMateReverseComplemented() ? -1 : 1
  }

  name() {
    return this.get('_read_name')
  }

  _read_name() {
    const nl = this.get('_l_read_name')
    const { byteArray, start } = this.bytes
    const decoder = new TextDecoder('ascii')
    return decoder.decode(byteArray.subarray(start + 36, start + 36 + nl - 1))
  }

  /**
   * Get the value of a tag, parsing the tags as far as necessary.
   * Only called if we have not already parsed that field.
   */
  _parseTag(tagName?: string) {
    // if all of the tags have been parsed and we're still being
    // called, we already know that we have no such tag, because
    // it would already have been cached.
    if (this._allTagsParsed) {
      return
    }

    const { byteArray, start } = this.bytes
    let p =
      this._tagOffset ||
      start +
        36 +
        this.get('_l_read_name') +
        this.get('_n_cigar_op') * 4 +
        this.get('_seq_bytes') +
        this.get('seq_length')

    const blockEnd = this.bytes.end
    let lcTag
    while (p < blockEnd && lcTag !== tagName) {
      const tag = String.fromCharCode(byteArray[p], byteArray[p + 1])
      lcTag = tag.toLowerCase()
      const type = String.fromCharCode(byteArray[p + 2])
      p += 3

      let value
      switch (type) {
        case 'A': {
          value = String.fromCharCode(byteArray[p])
          p += 1
          break
        }
        case 'i': {
          value = this.dv.getInt32(p, true)
          p += 4
          break
        }
        case 'I': {
          value = this.dv.getUint32(p, true)
          p += 4
          break
        }
        case 'c': {
          value = this.dv.getInt8(p)
          p += 1
          break
        }
        case 'C': {
          value = this.dv.getUint8(p)
          p += 1
          break
        }
        case 's': {
          value = this.dv.getInt16(p, true)
          p += 2
          break
        }
        case 'S': {
          value = this.dv.getUint16(p, true)
          p += 2
          break
        }
        case 'f': {
          value = this.dv.getFloat32(p, true)
          p += 4
          break
        }
        case 'Z':
        case 'H': {
          value = ''
          while (p <= blockEnd) {
            const cc = byteArray[p++]
            if (cc === 0) {
              break
            } else {
              value += String.fromCharCode(cc)
            }
          }
          break
        }
        case 'B': {
          value = ''
          const cc = byteArray[p++]
          const Btype = String.fromCharCode(cc)
          const limit = this.dv.getInt32(p, true)
          p += 4
          if (Btype === 'i') {
            if (tag === 'CG') {
              for (let k = 0; k < limit; k++) {
                const cigop = this.dv.getUint32(p, true)
                const lop = cigop >> 4
                const op = CIGAR_DECODER[cigop & 0xf]
                value += lop + op
                p += 4
              }
            } else {
              for (let k = 0; k < limit; k++) {
                value += this.dv.getInt32(p, true)
                if (k + 1 < limit) {
                  value += ','
                }
                p += 4
              }
            }
          }
          if (Btype === 'I') {
            if (tag === 'CG') {
              for (let k = 0; k < limit; k++) {
                const cigop = this.dv.getInt32(p, true)
                const lop = cigop >> 4
                const op = CIGAR_DECODER[cigop & 0xf]
                value += lop + op
                p += 4
              }
            } else {
              for (let k = 0; k < limit; k++) {
                value += this.dv.getUint32(p, true)
                if (k + 1 < limit) {
                  value += ','
                }
                p += 4
              }
            }
          }
          if (Btype === 's') {
            for (let k = 0; k < limit; k++) {
              value += this.dv.getInt16(p, true)
              if (k + 1 < limit) {
                value += ','
              }
              p += 2
            }
          }
          if (Btype === 'S') {
            for (let k = 0; k < limit; k++) {
              value += this.dv.getUint16(p, true)
              if (k + 1 < limit) {
                value += ','
              }
              p += 2
            }
          }
          if (Btype === 'c') {
            for (let k = 0; k < limit; k++) {
              value += this.dv.getInt8(p)
              if (k + 1 < limit) {
                value += ','
              }
              p += 1
            }
          }
          if (Btype === 'C') {
            for (let k = 0; k < limit; k++) {
              value += this.dv.getUint8(p)
              if (k + 1 < limit) {
                value += ','
              }
              p += 1
            }
          }
          if (Btype === 'f') {
            for (let k = 0; k < limit; k++) {
              value += this.dv.getFloat32(p, true)
              if (k + 1 < limit) {
                value += ','
              }
              p += 4
            }
          }
          break
        }
        default: {
          console.warn(`Unknown BAM tag type '${type}', tags may be incomplete`)
          value = undefined
          p = blockEnd
        } // stop parsing tags
      }

      this._tagOffset = p

      this._tagList.push(tag)
      if (lcTag === tagName) {
        return value
      }

      this.data[lcTag] = value
    }
    this._allTagsParsed = true
    return
  }

  _parseAllTags() {
    this._parseTag('')
  }

  _parseCigar(cigar: string) {
    return (
      //@ts-expect-error
      cigar
        .match(/\d+\D/g)
        //@ts-expect-error
        .map(op => [op.match(/\D/)[0].toUpperCase(), Number.parseInt(op, 10)])
    )
  }

  /**
   * @returns {boolean} true if the read is paired, regardless of whether both segments are mapped
   */
  isPaired() {
    return !!(this.flags & Constants.BAM_FPAIRED)
  }

  /** @returns {boolean} true if the read is paired, and both segments are mapped */
  isProperlyPaired() {
    return !!(this.flags & Constants.BAM_FPROPER_PAIR)
  }

  /** @returns {boolean} true if the read itself is unmapped; conflictive with isProperlyPaired */
  isSegmentUnmapped() {
    return !!(this.flags & Constants.BAM_FUNMAP)
  }

  /** @returns {boolean} true if the read itself is unmapped; conflictive with isProperlyPaired */
  isMateUnmapped() {
    return !!(this.flags & Constants.BAM_FMUNMAP)
  }

  /** @returns {boolean} true if the read is mapped to the reverse strand */
  isReverseComplemented() {
    return !!(this.flags & Constants.BAM_FREVERSE)
  }

  /** @returns {boolean} true if the mate is mapped to the reverse strand */
  isMateReverseComplemented() {
    return !!(this.flags & Constants.BAM_FMREVERSE)
  }

  /** @returns {boolean} true if this is read number 1 in a pair */
  isRead1() {
    return !!(this.flags & Constants.BAM_FREAD1)
  }

  /** @returns {boolean} true if this is read number 2 in a pair */
  isRead2() {
    return !!(this.flags & Constants.BAM_FREAD2)
  }

  /** @returns {boolean} true if this is a secondary alignment */
  isSecondary() {
    return !!(this.flags & Constants.BAM_FSECONDARY)
  }

  /** @returns {boolean} true if this read has failed QC checks */
  isFailedQc() {
    return !!(this.flags & Constants.BAM_FQCFAIL)
  }

  /** @returns {boolean} true if the read is an optical or PCR duplicate */
  isDuplicate() {
    return !!(this.flags & Constants.BAM_FDUP)
  }

  /** @returns {boolean} true if this is a supplementary alignment */
  isSupplementary() {
    return !!(this.flags & Constants.BAM_FSUPPLEMENTARY)
  }

  cigar() {
    if (this.isSegmentUnmapped()) {
      return
    }

    const { byteArray, start } = this.bytes
    const numCigarOps = this.get('_n_cigar_op')
    let p = start + 36 + this.get('_l_read_name')
    const seqLen = this.get('seq_length')
    let cigar = ''
    let lref = 0

    // check for CG tag by inspecting whether the CIGAR field
    // contains a clip that consumes entire seqLen
    let cigop = this.dv.getInt32(p, true)
    let lop = cigop >> 4
    let op = CIGAR_DECODER[cigop & 0xf]
    if (op === 'S' && lop === seqLen) {
      // if there is a CG the second CIGAR field will
      // be a N tag the represents the length on ref
      p += 4
      cigop = this.dv.getInt32(p, true)
      lop = cigop >> 4
      op = CIGAR_DECODER[cigop & 0xf]
      if (op !== 'N') {
        console.warn('CG tag with no N tag')
      }
      this.data.length_on_ref = lop
      return this.get('CG')
    } else {
      for (let c = 0; c < numCigarOps; ++c) {
        cigop = this.dv.getInt32(p, true)
        lop = cigop >> 4
        op = CIGAR_DECODER[cigop & 0xf]
        cigar += lop + op

        // soft clip, hard clip, and insertion don't count toward
        // the length on the reference
        if (op !== 'H' && op !== 'S' && op !== 'I') {
          lref += lop
        }

        p += 4
      }

      this.data.length_on_ref = lref
      return cigar
    }
  }

  _flags() {}

  length_on_ref() {
    if (this.data.length_on_ref) {
      return this.data.length_on_ref
    } else {
      this.get('cigar') // the length_on_ref is set as a side effect
      return this.data.length_on_ref
    }
  }

  _n_cigar_op() {
    return this.get('_flag_nc') & 0xffff
  }

  _l_read_name() {
    return this.get('_bin_mq_nl') & 0xff
  }

  /**
   * number of bytes in the sequence field
   */
  _seq_bytes() {
    return (this.get('seq_length') + 1) >> 1
  }

  getReadBases() {
    return this.seq()
  }

  seq() {
    const { byteArray, start } = this.bytes
    const p =
      start + 36 + this.get('_l_read_name') + this.get('_n_cigar_op') * 4
    const seqBytes = this.get('_seq_bytes')
    const len = this.get('seq_length')
    let buf = ''
    let i = 0
    for (let j = 0; j < seqBytes; ++j) {
      const sb = byteArray[p + j]
      buf += SEQRET_DECODER[(sb & 0xf0) >> 4]
      i++
      if (i < len) {
        buf += SEQRET_DECODER[sb & 0x0f]
        i++
      }
    }
    return buf
  }

  // adapted from igv.js
  getPairOrientation() {
    if (
      !this.isSegmentUnmapped() &&
      !this.isMateUnmapped() &&
      this._refID === this._next_refid()
    ) {
      const s1 = this.isReverseComplemented() ? 'R' : 'F'
      const s2 = this.isMateReverseComplemented() ? 'R' : 'F'
      let o1 = ' '
      let o2 = ' '
      if (this.isRead1()) {
        o1 = '1'
        o2 = '2'
      } else if (this.isRead2()) {
        o1 = '2'
        o2 = '1'
      }

      const tmp = []
      const isize = this.template_length()
      if (isize > 0) {
        tmp[0] = s1
        tmp[1] = o1
        tmp[2] = s2
        tmp[3] = o2
      } else {
        tmp[2] = s1
        tmp[3] = o1
        tmp[0] = s2
        tmp[1] = o2
      }
      return tmp.join('')
    }
    return ''
  }

  _bin_mq_nl() {
    return this.dv.getInt32(this.bytes.start + 12, true)
  }

  _flag_nc() {
    return this.dv.getInt32(this.bytes.start + 16, true)
  }

  seq_length() {
    return this.dv.getInt32(this.bytes.start + 20, true)
  }

  _next_refid() {
    return this.dv.getInt32(this.bytes.start + 24, true)
  }

  _next_pos() {
    return this.dv.getInt32(this.bytes.start + 28, true)
  }

  template_length() {
    return this.dv.getInt32(this.bytes.start + 32, true)
  }

  toJSON() {
    const data: { [key: string]: any } = {}
    for (const k of Object.keys(this)) {
      if (k.charAt(0) === '_' || k === 'bytes') {
        continue
      }
      //@ts-expect-error
      data[k] = this[k]
    }

    return data
  }
}
